### ISSUE TITLE
[Question] How to create WordPress datamodel?

### ISSUE THREAD

**Issue by cr101** [prisma/prisma2#324]
I'm using a MySQL database to store different content types (i.e companies, jobs, events, etc) which are all linked to each other and all have tags and hierarchical categories.

The structure of my database is very similar to the WordPress database, however, all the meta fields for each content type are stored in a separate *_Meta table instead of one single meta table.

Is it possible to do this in Prisma 2? If so, how should I define my datamodel so I can do a query like below?
![Screenshot_7](https://user-images.githubusercontent.com/4064582/62460278-b559f780-b7d5-11e9-9390-ef1ddc48cd7f.jpg)

Below is my data model definition:

```groovy
enum ContentStatus {
  ACTIVE
  DRAFT
  PENDING
  DELETED
 }
 
enum ContentType {
  COMPANY  
  JOB
  EVENT  
 }

model User {
  id: Int @id
  name: String!
  email: String @unique
  password: String!
  createdAt DateTime @default(now())
}

model Content {
  id: Int @id @unique
  title: String!
  description: String!
  excerpt: String!
  status: ContentStatus! @default(ACTIVE)
  slug: String @unique
  type: ContentType! @default(COMPANY)
  createdBy: User
  createdAt:  DateTime   @default(now())
  updatedAt:  DateTime   @updatedAt
}

model CompanyMeta {
  id: Int @id @unique
  content: Content
  events: Content! @relation(link: TABLE, name: "CompanyEventRelationships")
  incorporated: DateTime!
  email: String
  phone: String
  in_url: String
  fb_url: String
}

model EventMeta {
  id: Int @id @unique
  content: Content
  company: Content! @relation(link: TABLE, name: "CompanyEventRelationships")
  startDate: DateTime
  endDate: DateTime
  price: Int
  registrationUrl: String
}

model CompanyEventRelationships @relationTable {
  id: Int @id
  company: Content
  event: Content
}

enum TaxonomyType {
    CATEGORY
    TAG
 }

model Term {
  id: Int @id @unique
  name: String
  slug: String @unique
  term_group: Int @default(value: 0)
}

model TermTaxonomy {
  id: ID! @unique
  term: Term! @relation(link: INLINE) @default(value: 0)
  taxonomy: TaxonomyType!
  description: String?
  parent: Term! @relation(link: INLINE) @default(value: 0)
  count: Int @default(value: 0)
}

model TermRelationship {
  content: Content! @relation(link: INLINE) @default(value: 0)
  term_taxonomy: TermTaxonomy! @relation(link: INLINE) @default(value: 0)
  order: Int @default(value: 0)
}
```

**Comment by pantharshit00** [comment]
(edited your datamodel to get syntax highlighting

First of all, you will need to implement that GraphQL API yourself as Prisma 2 only concerns with data access.

For data modelling, I think also need to define an operational relation between the types, so they can be connected via a foreign key. In the backend, you can resolve using the type field.

```
model Content {
  id: Int @id @unique
  title: String
  description: String
  excerpt: String
  status: ContentStatus @default(ACTIVE)
  slug: String @unique
  type: ContentType @default(COMPANY)
  createdBy: User
  companyMeta: CompanyMeta?
  # .... other metas.
  createdAt:  DateTime   @default(now())
  updatedAt:  DateTime   @updatedAt
}
```
Also, everything is required by default in the new syntax, you use `?` to make things optional so your model has some syntax errors.


Also, you can introspect a WordPress instance using our CLI to see the model that would have used to get an idea.

**Comment by cr101** [comment]
@pantharshit00 thanks for your help.

How do I use a relation table to link companies and events together to reflect a many-to-many relationship? PS: I have modified my datamodel to show what I'm trying to achieve.

How do I create the connection between the content type and the hierarchical categories?

**Comment by pantharshit00** [comment]
Many to many relationships are defined using the following syntax:

```

model A {
  b B[]
}

model B {
 a A[]
}
```

You can also customize the join tables, here is part of the spec which deals with many to many relationships: https://github.com/prisma/specs/tree/master/prisma-schema-language#explicit-many-to-many-mn-relationships


Also, the relation between category and content-type will set by a foreign key:

```
model Category  {
 type ContentType
}

model ContentType {
 categories Category
}
```

**Comment by cr101** [comment]
@pantharshit00  The tags and categories for all the ContentTypes are stored in 3 database tables (Term, TermTaxonomy, TermRelationship) just like in the [WordPress database](https://codex.wordpress.org/Database_Description). How do I create a relation between content-types and categories, tags?

**Comment by pantharshit00** [comment]
I went ahead and introspected a wordpress instance for you, this should give you a good idea:
```groovy
datasource db {
  provider = "mysql"
  url      = "mysql://exampleuser:examplepass@localhost:2234/exampledb"
}

model WpComment {
  comment_ID         Int      @id
  commentAgent       String   @map("comment_agent")
  commentApproved    String   @map("comment_approved")
  commentAuthorEmail String   @map("comment_author_email")
  commentAuthorIP    String   @map("comment_author_IP")
  commentAuthorUrl   String   @map("comment_author_url")
  commentContent     String   @map("comment_content")
  commentDate        DateTime @map("comment_date")
  commentDateGmt     DateTime @map("comment_date_gmt")
  commentKarma       Int      @map("comment_karma")
  commentParent      Int      @map("comment_parent")
  commentPostID      Int      @map("comment_post_ID")
  commentType        String   @map("comment_type")
  userId             Int      @map("user_id")

  @@map("wp_comments")
}

model WpCommentmeta {
  meta_id   Int     @id
  commentId Int     @map("comment_id")
  metaKey   String? @map("meta_key")
  metaValue String? @map("meta_value")

  @@map("wp_commentmeta")
}

model WpLink {
  link_id         Int      @id
  linkDescription String   @map("link_description")
  linkImage       String   @map("link_image")
  linkName        String   @map("link_name")
  linkNotes       String   @map("link_notes")
  linkOwner       Int      @map("link_owner")
  linkRating      Int      @map("link_rating")
  linkRel         String   @map("link_rel")
  linkRss         String   @map("link_rss")
  linkTarget      String   @map("link_target")
  linkUpdated     DateTime @map("link_updated")
  linkUrl         String   @map("link_url")
  linkVisible     String   @map("link_visible")

  @@map("wp_links")
}

model WpOption {
  option_id   Int    @id
  autoload    String
  optionName  String @map("option_name") @unique
  optionValue String @map("option_value")

  @@map("wp_options")
}

model WpPost {
  ID                  Int      @id
  commentCount        Int      @map("comment_count")
  commentStatus       String   @map("comment_status")
  guid                String
  menuOrder           Int      @map("menu_order")
  pinged              String
  pingStatus          String   @map("ping_status")
  postAuthor          Int      @map("post_author")
  postContent         String   @map("post_content")
  postContentFiltered String   @map("post_content_filtered")
  postDate            DateTime @map("post_date")
  postDateGmt         DateTime @map("post_date_gmt")
  postExcerpt         String   @map("post_excerpt")
  postMimeType        String   @map("post_mime_type")
  postModified        DateTime @map("post_modified")
  postModifiedGmt     DateTime @map("post_modified_gmt")
  postName            String   @map("post_name")
  postParent          Int      @map("post_parent")
  postPassword        String   @map("post_password")
  postStatus          String   @map("post_status")
  postTitle           String   @map("post_title")
  postType            String   @map("post_type")
  toPing              String   @map("to_ping")

  @@map("wp_posts")
}

model WpPostmeta {
  meta_id   Int     @id
  metaKey   String? @map("meta_key")
  metaValue String? @map("meta_value")
  postId    Int     @map("post_id")

  @@map("wp_postmeta")
}

model WpTerm {
  term_id   Int    @id
  name      String
  slug      String
  termGroup Int    @map("term_group")

  @@map("wp_terms")
}

model WpTermRelationship {
  /// Multiple ID fields (compound indexes) are not supported
  object_id        Int @id
  /// Multiple ID fields (compound indexes) are not supported
  term_taxonomy_id Int @id
  termOrder        Int @map("term_order")

  @@map("wp_term_relationships")
}

model WpTermTaxonomy {
  term_taxonomy_id Int    @id
  count            Int
  description      String
  parent           Int
  taxonomy         String
  termId           Int    @map("term_id")

  @@map("wp_term_taxonomy")
}

model WpTermmeta {
  meta_id   Int     @id
  metaKey   String? @map("meta_key")
  metaValue String? @map("meta_value")
  termId    Int     @map("term_id")

  @@map("wp_termmeta")
}

model WpUser {
  ID                Int      @id
  displayName       String   @map("display_name")
  userActivationKey String   @map("user_activation_key")
  userEmail         String   @map("user_email")
  userLogin         String   @map("user_login")
  userNicename      String   @map("user_nicename")
  userPass          String   @map("user_pass")
  userRegistered    DateTime @map("user_registered")
  userStatus        Int      @map("user_status")
  userUrl           String   @map("user_url")

  @@map("wp_users")
}

model WpUsermeta {
  umeta_id  Int     @id
  metaKey   String? @map("meta_key")
  metaValue String? @map("meta_value")
  userId    Int     @map("user_id")

  @@map("wp_usermeta")
}
```

**Comment by cr101** [comment]
@pantharshit00 Thank you for doing the WordPress database introspection. I'm still not sure how to create the relations between posts and categories, tags so I can do queries like below:

**Get list of Tags**
```groovy
query GET_TAGS {
  tags {
    edges {
      node {
        id
        tagId
        name
      }
    }
  }
}
```
**Query a list of Category nodes**
```groovy
query GET_CATEGORIES {
  categories {
    edges {
      node {
        id
        categoryId
        name
      }
    }
  }
}
```
**Get a list of Tags, and on each Tag, a list of posts it's connected to**
```groovy
query GET_TAGS {
  tags {
    edges {
      node {
        id
        tagId
        name
        posts {
          nodes {
            id
            title
            date
          }
        }
      }
    }
  }
}
```
**Get Hierarchical Taxonomies**
```groovy
query GET_CATEGORIES {
  categories {
    edges {
      node {
        id
        categoryId
        name
        children {
          nodes {
            id
            name
            children {
              nodes {
                id
                name
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by pantharshit00** [comment]
I think you are considering Prisma to be a GraphQL ORM which it is not. 

You will need to implement these in your GraphQL server.


But I do think we can provide an example for this once the Relay style pagination part of the photon is ready(https://github.com/prisma/specs/blob/photonjs-improvements/specs/PhotonJS.md#todo-withpageinfo)

**Comment by cr101** [comment]
@pantharshit00 

> You will need to implement these in your GraphQL server.
Yes, the plan is to use `nexus-prisma` and `apollo-server`

An example would be very helpful for me and many other developers.

**Comment by janpio** [comment]
@pantharshit00 Can you please spin out the `kind/docs` issue here into its own issue with its own description? Thanks.

**Comment by owcc** [comment]
> I think you are considering Prisma to be a GraphQL ORM which it is not.
> 
> You will need to implement these in your GraphQL server.
> 
> But I do think we can provide an example for this once the Relay style pagination part of the photon is ready(https://github.com/prisma/specs/blob/photonjs-improvements/specs/PhotonJS.md#todo-withpageinfo)

Url is lost, any photon example with relay style pagination?

**Comment by pantharshit00** [comment]
@owcc Here is the new photon spec: https://github.com/prisma/specs/tree/master/photonjs

**Comment by divyenduz** [comment]
Wordpress datamodel can be created by introspecting it using the prisma2 introspect command. This however would run into this issue right now: https://github.com/prisma/prisma2/issues/1113

Please track that issue, and once that is resolved, please try to introspect an existing wordpress datamodel via `prisma2 introspect --url <connection string>`

**Comment by jasonduffus** [comment]
fix this prisma schema class to represent a facet hierarchical relationship structure.

model Taxonomy {
  id        String       @id @default(uuid())
  name      String
  slug      String
  parent    String?
  type      TaxonomyType @map("TaxonomyType")
  appliesTo String
  products  ProductTag[]
  posts     PostTag[]
}
